parlai/tasks/cmu_dog/agents.py

```diff
@@ -0,0 +1,364 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+import copy
+import json
+import os
+from enum import Enum
+from typing import List, Optional, Tuple
+
+from parlai.core.message import Message
+from parlai.core.metrics import F1Metric
+from parlai.core.mutators import EpisodeMutator, MessageMutator, register_mutator
+from parlai.core.opt import Opt
+from parlai.core.params import ParlaiParser
+from parlai.core.teachers import DialogTeacher
+from parlai.tasks.cmu_dog.build import build
+from parlai.tasks.wizard_of_wikipedia.agents import (
+    RareWordF1Calculator,
+    TOKEN_KNOWLEDGE,
+    TOKEN_END_KNOWLEDGE,
+)
+from parlai.utils.data import DatatypeHelper
+from parlai.utils.io import PathManager
+from parlai.utils.logging import logger
+from parlai.utils.typing import TShared
+
+
+SILENCE = '__SILENCE__'
+
+
+class SplitType(Enum):
+    ORIGINAL = "original"
+    ORIGINAL_DEDUPED = "deduped"
+    SEEN = "seen"
+    UNSEEN = "unseen"
+
+
+def _datapath(opt: Opt) -> str:
+    build(opt)
+    return os.path.join(opt['datapath'], 'cmu_dog')
+
+
+def _datafile(split: str, split_type: SplitType) -> str:
+    """
+    Returns the filename, e.g. train.json.
+    """
+    if split_type == SplitType.ORIGINAL:
+        return f"{split}.json"
+    if split_type == SplitType.SEEN:
+        if 'test' in split:
+            return "test_seen_split_seen_unseen.json"
+        return f"{split}_split_seen_unseen.json"
+    if split_type == SplitType.UNSEEN:
+        if "test" not in split:
+            logger.warning(
+                "Trying to use a non-test dataset with split `unseen`. `unseen` "
+                "only returns the unseen test set. Are you sure you didn't mean to "
+                "use `seen` here?"
+            )
+        return "test_unseen_split_seen_unseen.json"
+    return f"{split}_deduped.json"
+
+
+def _all_split_datafiles(opt: Opt) -> List[str]:
+    datafiles = []
+    split_type = SplitType(opt.get("cmu_dog_split_type"))
+    if split_type in {SplitType.SEEN, SplitType.UNSEEN}:
+        # For seen/unseen split, the full set of dialogs is split
+        # across train, valid, test seen, and test unseen
+        for split in ['train', 'valid', 'test']:
+            datafiles.append(_datafile(split, SplitType.SEEN))
```

Comment:
[No-op, just a comment] As a sidenote, 

> list(train_seen + valid_seen + test_seen + test_unseen) = list(train_original_deduped + valid_original_deduped + test_original_deduped)

so we could do that... but the code is probably clearer this way. 

(On that note, this does mean that we will be duplicate counting conversations when we do F1 on the original split, but it seems like we're deciding to be okay with that.)

Reply:
> list(train_seen + valid_seen + test_seen + test_unseen) = list(train_original_deduped + valid_original_deduped + test_original_deduped)

I'm leaning toward leaving these separate for code clarity. But it might be good to add this relationship as a unit test so that it's documented.

> we will be duplicate counting conversations when we do F1 on the original split

When you say "original split" you mean the one that isn't deduplicated, right? I guess I'm kind of assuming that if someone is intentionally using the split with duplicate conversations, it's probably desirable that the conversations be duplicated in the metric calculations as well.

Maybe I should add a warning print if people use the `original` split type, and point them to your issue (https://github.com/festvox/datasets-CMU_DoG/issues/2) so they know what they're getting into.